Feature/date filter

dylanirlbeck opened the pull request:
Fixes #48

~This PR is a WIP -- all the features work and a basic design is in place, but I wanted to put it up for feedback in case anyone is curious and can provide some early feedback. The code is pretty ugly, so I plan on cleaning it up.~

This PR adds an option for filtering the generated artwork by a specified date range. I also added some custom validation, to prevent users from entering dates less than `-8000` (8000 BCE), greater than `2020` (the current year), or entering a `From` date greater than a `To` date.

In addition, I noticed this project is not using a standard formatter for JS/HTML/CSS, so I went ahead and used [Prettier](https://prettier.io/). Most of the diff will likely be associated with this reformatting, and won't be actual functional changes.

## UI

Note, I'm not a designer. Feedback on the UI is much appreciated! At the moment the form has some basic validations (you'll get a red border around the input and be prevented from submitting) for the following conditions:
* The `dateRangeFrom` is greater than the `dateRangeTo`
* The `dateRangeFrom` or `dateRangeTo` are less than `-8000`
* The `dateRangeFrom` or `dateRangeTo` are greater than the current year* (Note to self: need to add this one)

<img width="762" alt="image" src="https://user-images.githubusercontent.com/35497479/86517553-7f5f3200-bdef-11ea-8fe6-370f58d66994.png">

surreal8 commented:
This is great @dylanirlbeck! Thank you! I like that the setting is incorporated into the extension's settings, and not directly on the page. Design/UX wise I think it looks nice and clean, only thing is we probably don't need the scroll bars for this one item, and when you hit 'ok' I'd prefer if the settings box closed, but we should run all this by our designer when she has a moment. 

I put a date range between 1995-2020 and I got this 1 artwork, https://www.artic.edu/artworks/128480/arch-of-pantani-plate-thirty-seven-from-the-ruins-of-rome?utm_medium=chrome-extension&utm_source=Arch%20of%20Pantani,%20plate%20thirty-seven%20from%20the%20Ruins%20of%20Rome, whose date data doesn't seem to be correct, right @IllyaMoskvin and @nikhiltri? 

1980-2020 yields 3 artworks :).

dylanirlbeck commented:
@surreal8 Thanks for taking a look! I second the call for the `OK` button to close the modal -- I'll see if I can get that to work. I'm not sure the best way to indicate to the user that the options have been saved correctly, but I'm sure that's something we can talk through before merging in to `master`. 

> only thing is we probably don't need the scroll bars for this one item,

Can you clarify what you mean here? I'm not seeing scroll bars on my options modal, though I'm not sure if this is what you mean. 

> I put a date range between 1995-2020 and I got this 1 artwork

That happened to me as well. I'll try and debug the code to see if there's something going wrong on the extension end (as opposed to Elasticsearch).

dylanirlbeck commented on options.css:
```diff
@@ -0,0 +1,130 @@
+#filter-footer {
```
See which CSS I can remove -- I suspect there's quite a few redundant styles.

dylanirlbeck commented on options.js:
```diff
@@ -0,0 +1,84 @@
+// Restores select box and checkbox state using the preferences
+// stored in chrome.storage.
+function restore_options() {
+    // Use default value color = 'red' and likesColor = true.
```
Remove this comment

dylanirlbeck commented on options.html:
```diff
@@ -0,0 +1,44 @@
+                    <input
+                        type="number"
+                        min="-8000"
+                        placeholder="Min: e.g. -8000"
```
Change placeholders to be `From` and `To` as opposed to `Min` and `Max`

IllyaMoskvin commented on script.js:
```diff
@@ -48,26 +47,39 @@
         // ...returns `null` if not found. JSON.parsing `null` also returns `null`
-        let savedResponse = JSON.parse(localStorage.getItem(savedResponseKey));
+        //let savedResponse = JSON.parse(localStorage.getItem(savedResponseKey));
+        savedResponse = null;
```
Ah, I understand that this is being done to cache-bust the saved response, but we need to keep this functionality for performance reasons. This makes it so that the new tab doesn't have to perform an API request before showing the image, which saves about 200 ms.

As an alternative, maybe there is a way to unset the saved response in `save_options`?

dylanirlbeck replied:
Gotcha, good call. I modified the logic slightly to compare the current and saved options to see if they differ, and if so run the new response. This seems to maintain good performance!

IllyaMoskvin commented on options.css:
```diff
@@ -0,0 +1,82 @@
+    font-family: "Ideal Sans A", "Ideal Sans B", Helvetica, Arial, sans-serif;
+}
+
+.range-slider__custom-row {
```
Given that the scope changed to make this not a range slider for the time being, should these classes be renamed?

IllyaMoskvin commented on options.css:
```diff
@@ -0,0 +1,82 @@
+
+#filter-status {
+    font-size: 14px;
+    font-family: "Ideal Sans A", "Ideal Sans B", Helvetica, Arial, sans-serif;
```
Although we use Ideal Sans on the website, we can't use it here for licensing reasons. Let's use Open Sans instead. Our designer recommended it as the alternative to Ideal Sans for the context. We already use it for the new tab itself:

https://github.com/art-institute-of-chicago/browser-extension/blob/bc41582ed38398423aa9ecc564230b365f46386a/index.html#L6

We will for sure check in with her about the options page once we've got more filters implemented, but for now, let's use Amiri for headlines and Open Sans for everything else.

dylanirlbeck commented on options.js:
```diff
@@ -0,0 +1,76 @@
+    );
+}
+
+function validateDateInputs() {
```
Thanks for the helpful advice! I had never used in-browser validation before (React abstracted all that away from me), but I was able to get it working.

IllyaMoskvin replied:
Nice! Sure thing. The cross-browser support is [surprisingly good](https://caniuse.com/#feat=form-validation). With the browser extension, where we have a pretty good idea about what browser the user will have, I think we can safely rely on it.

dylanirlbeck commented on manifest.json:
```diff
@@ -12,9 +16,7 @@
-    "permissions": [
-        "activeTab"
-    ],
+    "permissions": ["activeTab", "storage"],
```
Is there an option in the prettier config for this? I don't think I can add a `//prettier-ignore` in a JSON file.

IllyaMoskvin replied:
I wrote my comment before learning more about Prettier. We could add `manifest.json` to `.prettierignore`, but honestly, I'm inclined to not worry about it. I think this is just fine as-is.

dylanirlbeck commented on manifest.json:
```diff
@@ -4,6 +4,10 @@
     "author": "Art Institute of Chicago",
+    "options_ui": {
+        "page": "options.html",
+        "open_in_tab": false
```
I'm open to either, honestly -- having the extra space will probably be good for scalability purposes, so I'll change the boolean to `true` for now.

dylanirlbeck commented on options.js:
```diff
@@ -0,0 +1,76 @@
+// Saves options to chrome.storage
+function save_options() {
+    var dateRangeFrom = document.getElementById("dateRangeFrom").value;
+    var dateRangeTo = document.getElementById("dateRangeTo").value;
```
Hmm, can you take a screenshot/GIF showing me what you mean? I modified these calls to use short-circuiting (i.e. `stuff.value || null`), which _might_ address your concern, but I'm more unsure than not.

IllyaMoskvin replied:
Before you added short-circuiting, if I saved the options form while those two fields were empty, they got saved as `""`, and `chrome.storage.sync.get` retrieved them as empty strings, which then got passed to `getQuery`, where they were converted by the `Number()` call from `""` to `0`.

After the change, they got saved as `null` instead of `""`, but it seems that `chrome.storage.sync.get` does not use the default values for values that were saved as `null`... maybe the default values only get used for undefined options? I'm honestly not yet clear on how it the defaults work.

I changed `options.js` to make it so that if these two fields are empty, we save the default values instead of `null`. This seems to have fixed the issue.

dylanirlbeck commented:
@IllyaMoskvin My apologies for the delay, and I hope all is well with you and the ARTIC! I was actually able to visit last weekend -- l especially liked Malangatana 😄 

 My internship got very busy, and I was moving all over the place. I believe I've addressed all your comments in the original review, so I think this is ready for you to take another look!

One thing to note: in order to minimize the diff I changed the base branch from `feature/settings` to `master`. If you could go in and rebase `feature/settings` against `master` to get the Prettier changes it'd be much appreciated! I can then modify the base to point back to `feature/settings`.
